src/SSTIOImp.hpp

```diff
@@ -0,0 +1,217 @@
+            };
+
+            struct sst_mmio_interface_s
+            {
+                uint32_t is_write;
+                uint32_t cpu_index;
+                uint32_t register_offset;
+                uint32_t value;
+            };
+
+            struct sst_mmio_interface_batch_s
+            {
+                uint32_t num_entries;
+                sst_mmio_interface_s interfaces[1];
+            };
+
+            struct sst_mbox_interface_s
+            {
+                uint32_t cpu_index;
+                uint32_t mbox_interface_param; // Parameter to the mbox interface itself
+                uint32_t write_value; // Mailbox data, or input parameter for a read
+                uint32_t read_value; // Mailbox data (read-only)
+                uint16_t command;
+                uint16_t subcommand;
+                uint32_t reserved;
+            };
+
+            struct sst_mbox_interface_batch_s
+            {
+                uint32_t num_entries;
+                sst_mbox_interface_s interfaces[1];  // TODO: might need to be array instead
+            };
+
+            template<typename ListStruct, typename CommandStruct>
+            class CommandList
+            {
+                public:
+                explicit CommandList(size_t command_count);
+            };
+
+            template<typename OuterStruct>
+            using InnerStruct =
+                typename std::remove_all_extents<decltype(OuterStruct::interfaces)>::type;
+
+            // Given a single vector of messages to send to an ioctl, split it
+            // into multiple structs to send to that ioctl. Each InnerStruct
+            // contains a single message. Each OuterStruct contains multiple
+            // messages, with size upper-bounded by m_batch_command_limit.
+            template<typename OuterStruct>
+            std::vector<std::unique_ptr<OuterStruct> >
+                ioctl_structs_from_vector(const std::vector<InnerStruct<OuterStruct> >& commands)
+
+            {
+                std::vector<std::unique_ptr<OuterStruct> > outer_structs;
+
+                int handled_commands = 0;
+                while (handled_commands < commands.size())
+                {
+                    size_t batch_size = std::min(
+                        m_batch_command_limit,
+                        static_cast<int>(commands.size() - handled_commands));
+
+                    // The inner struct is embedded in the outer struct, and
+                    // the inner struct's size depends on how many entries it
+                    // can contain. That size is dynamically determined, so we
+                    // manually allocate the outer struct here.
+                    outer_structs.emplace_back(reinterpret_cast<OuterStruct *>(
+                        new char[sizeof(OuterStruct::num_entries) +
+                                 sizeof(InnerStruct<OuterStruct>) * commands.size()]));
+                    outer_structs.back()->num_entries = batch_size;
```

Comment:
in c++17 we can use the return value from emplace :)

Reply:
Things will just be too easy when we switch to C++17!